include/gul14/ThreadPool.h

```diff
@@ -0,0 +1,297 @@
+ *         pool.add_task([]() { std::cout << "Task 5, a second later\n"; }, 1s);
+ *     });
+ * \endcode
+ *
+ * All public member functions are thread-safe.
+ *
+ * On Linux, threads in the pool explicitly block the signals SIGALRM, SIGINT, SIGPIPE,
+ * SIGTERM, SIGURG, SIGUSR1, and SIGUSR2. This is done to prevent the threads from
+ * terminating the whole process if one of these signals is received.
+ *
+ * \since GUL version 2.11
+ */
+class ThreadPool
+{
+public:
+    using TaskId = std::uint64_t;
+    using TimePoint = std::chrono::time_point<std::chrono::system_clock>;
+    using Duration = TimePoint::duration;
+
+    /// Maximum possible capacity for the task queue.
+    constexpr static std::size_t max_capacity{ 10'000'000 };
+
+    /// Maximum possible number of threads
+    constexpr static std::size_t max_threads{ 10'000 };
+
+    /**
+     * Create a thread pool with the desired number of threads and the specified capacity
+     * for queuing tasks.
+     *
+     * This constructor launches the desired number of threads. The threads are joined
+     * when the ThreadPool object gets destroyed.
+     *
+     * \param num_threads  Desired number of threads
+     * \param capacity     Maximum number of pending tasks that can be queued
+     *
+     * \exception std::invalid_argument is thrown if the desired number of threads is
+     *            zero or greater than max_threads, or if the requested capacity is zero
+     *            or exceeds max_capacity.
+     */
+    ThreadPool(std::size_t num_threads, std::size_t capacity = 1000);
+
+    /**
+     * Destruct the ThreadPool and join all threads.
+     *
+     * This destructor blocks until all threads have terminated. Work that has not yet
+     * been started in one of the threads gets canceled, but work that has already been
+     * assigned to a thread continues to be executed until it completes.
+     */
+    ~ThreadPool();
+
+    /**
+     * Enqueue a task.
+     *
+     * There are multiple overloads of this function for variations of the arguments:
+     *
+     * \param fct   A function object or function pointer to be executed. This function
+     *              must return void and either takes no arguments (`void fct()`) or
+     *              a reference to the ThreadPool by which it gets executed
+     *              (`void fct(ThreadPool&)`). Null pointers are not allowed.
+     * \param start_time  Earliest time point at which the task is to be started
+     * \param name  Optional name for the task (mainly for debugging)
+     *
+     * \returns a unique ID to identify this task.
+     * \exception std::runtime_error is thrown if the queue is full. std::invalid_argument
+     *            is thrown if a null pointer is given.
+     */
+    TaskId add_task(std::function<void(ThreadPool&)> fct, TimePoint start_time = {},
+        std::string name = {});
+
+    TaskId add_task(std::function<void()> fct, TimePoint start_time = {},
```

Comment:
Ah, I forgot to mention this.

I think the returned thing (`TaskId`) is a bit underwhelming. Makes me feel like the 90s and `C` ;-)

I would expect an object, maybe something along this lines (apart from how-to-implement)

```c++
class SheduledTask {
    bool is_pending();
    bool is_running();
    bool is_finished();
    ThreadPool get_manager();
    // or rather
    std::optional<ThreadPool*> get_manager(); // ;-D
}

And while we are at it, if I want to load off some work to some other task, I usually want to get some data back, like with `std::future`. This whole thing smells a bit pre-std-future.  
Of course I could handle the return data somehow myself, but that is also error prone and can be easily (?) abstracted here once and for all.

I believe the tasks should be futures and not threads?

Reply:
And again this is a use-case for the `safe_self` ideom, where the `SheduledTask` holds a pointer to the `ThreadPool` that gets marked invalid on Threadpool destruction.

Reply:
I usually use `std::packaged_task` which is based on `std::future` which is I guess what we want.

This is related to [why put `perform_work` into a closure](https://github.com/gul-cpp/gul14/pull/62#discussion_r1475837184) question.

Reply:
I think returning something like a `ScheduledTask` (or `TaskHandle`?) is a good idea, especially if we can somehow manage to stick arbitrary return types of the functions into promises/futures. This will require a bit of work, but I could noodle on it in the next days or weeks.

Reply:
I guess that will be some effort ;-)
My suggestion would be to "keep everything as it is now" except that we return a `ScheduledTask` object instead of a handle number, and that can be expanded later without any/too much breakage.

And good for that object would be a way to access the associated ThreadPool object in a safe manner (see smart pointer juggling).
* Users of the return value of `add_task` have a more object oriented API
* Futures can be added to it later

Reply:
I have just pushed a new version that
- returns `TaskHandle` instead of a simple ID.
- `TaskHandle` has member functions `is_complete()`, `is_pending()`, `is_running()`.
- These member functions are safe against a deleted `ThreadPool`. If that is the case, they throw (but we could let them keep the pool alive instead).
- `TaskHandle` stores a future to a return value of arbitrary type. It can be retrieved with `get_result()`.

Keeping the references between the three classes intact is quite a nasty exercise ("smart pointer juggling"). The design of the `ThreadPool` was a lot clearer before that, but I admit that the resulting API is quite nice:
```cpp
ThreadPool pool(4);
auto task = pool.add_task([]() { ... });
while (not task.is_complete())
     display_progress();
auto result = task.get_result();
```

This is still in need of some cleanup and documentation, but the general concept is there.